Potential workflow/builders@4.0.1-beta.64 regression: Next.js build fails to resolve relative `.ts` imports in step files

### Description

After upgrading `workflow` from `4.2.0-beta.72` to `4.2.0-beta.73`, Next.js production builds fail with `Module not found` errors for relative TypeScript imports inside step files.

The regression is in the transitive dependency `@workflow/builders@4.0.1-beta.64`, which introduced ["Corrected dependency resolution for step imports with TypeScript extensions (.ts, .mts, .cts)"](https://github.com/vercel/workflow/releases). This "correction" breaks resolution of extensionless relative imports to `.ts` modules — the bundler looks for `.js` instead of `.ts`.

Is very strange, but seems that only happens in the vercel build not in local but not totally sure yet about this.

### Reproduction

A step file imports a local utility with a standard extensionless relative path:

```ts
// upload-video-and-build-output.step.ts
import { getVideoDimensions, getVideoFeatureCost, getVideoOriginAndFetchOptions } from "../utils/model-config";
```

The file `../utils/model-config.ts` exists. The build compiles workflow steps into `apps/web/app/.well-known/workflow/v1/step/route.js`, which fails:

```
Module not found: Can't resolve '../../../../../src/server/lib/tasks/video-clip/utils/model-config.js'
```

### Environment

- `workflow`: `4.2.0-beta.73`
- `@workflow/next`: `4.0.1-beta.69` (transitive)
- `@workflow/builders`: `4.0.1-beta.64` (transitive — **this is the breaking change**)
- `@workflow/rollup`: `4.0.0-beta.30` (transitive)
- `next`: `16.2.1`
- Node: `24`
- macOS

### Working versions

Everything builds correctly with `workflow@4.2.0-beta.72` and its transitive deps:
- `@workflow/builders@4.0.1-beta.63`
- `@workflow/next@4.0.1-beta.68`
- `@workflow/rollup@4.0.0-beta.29`

### Workaround

Pin `workflow` to `4.2.0-beta.72` or add pnpm overrides for the transitive deps:

```yaml
# pnpm-workspace.yaml
overrides:
  "@workflow/builders": 4.0.1-beta.63
  "@workflow/next": 4.0.1-beta.68
  "@workflow/rollup": 4.0.0-beta.29
```

### Related
- #1524  — suspected cause (not 100% certain)
- #1179 — same root cause (transitive local TS modules not bundled) but for Nuxt/Nitro
- #278 — original fix for direct `.ts` imports, which `beta.64` attempted to extend

--- a/.changeset/fix-ts-extension-rewrite-scope.md
+++ b/.changeset/fix-ts-extension-rewrite-scope.md
@@ -1,0 +1,5 @@
+---
+"@workflow/builders": patch
+---
+
+Only rewrite .ts extensions to .js in externalized step imports when targeting Node's native ESM loader (vitest), preserving original extensions for framework bundlers (Next.js, SvelteKit, etc.)

--- a/packages/builders/src/base-builder.ts
+++ b/packages/builders/src/base-builder.ts
@@ -337,6 +337,7 @@ export abstract class BaseBuilder {
     format = 'cjs',
     outfile,
     externalizeNonSteps,
+    rewriteTsExtensions,
     tsconfigPath,
     discoveredEntries,
   }: {
@@ -345,6 +346,7 @@ export abstract class BaseBuilder {
     outfile: string;
     format?: 'cjs' | 'esm';
     externalizeNonSteps?: boolean;
+    rewriteTsExtensions?: boolean;
     discoveredEntries?: DiscoveredEntries;
   }): Promise<{
     context: esbuild.BuildContext | undefined;
@@ -519,6 +521,7 @@ export abstract class BaseBuilder {
           outdir: outfile ? dirname(outfile) : undefined,
           projectRoot: this.transformProjectRoot,
           workflowManifest,
+          rewriteTsExtensions,
         }),
       ],
       // Plugin should catch most things, but this lets users hard override

--- a/packages/builders/src/swc-esbuild-plugin.test.ts
+++ b/packages/builders/src/swc-esbuild-plugin.test.ts
@@ -50,10 +50,47 @@ describe('createSwcPlugin externalizeNonSteps', () => {
     { inputExt: '.tsx', outputExt: '.js' },
     { inputExt: '.mts', outputExt: '.mjs' },
     { inputExt: '.cts', outputExt: '.cjs' },
-  ])('rewrites externalized $inputExt imports to $outputExt', async ({
+  ])('rewrites externalized $inputExt imports to $outputExt when rewriteTsExtensions is enabled', async ({
     inputExt,
     outputExt,
   }) => {
+    const outdir = join(testRoot, 'out');
+    const srcDir = join(testRoot, 'src');
+    const stepFile = join(srcDir, 'step.ts');
+
+    writeFile(join(srcDir, `dep${inputExt}`), 'export const dep = {};');
+    writeFile(stepFile, `import { dep } from './dep';\nconsole.log(dep);`);
+
+    const result = await esbuild.build({
+      entryPoints: [stepFile],
+      absWorkingDir: testRoot,
+      outdir,
+      bundle: true,
+      format: 'esm',
+      platform: 'node',
+      write: false,
+      plugins: [
+        createSwcPlugin({
+          mode: 'step',
+          entriesToBundle: [stepFile],
+          outdir,
+          rewriteTsExtensions: true,
+        }),
+      ],
+    });
+
+    expect(result.errors).toHaveLength(0);
+    const output = result.outputFiles[0].text;
+    expect(output).toContain(`/dep${outputExt}`);
+    expect(output).not.toContain(`/dep${inputExt}`);
+  });
+
+  it.each([
+    '.ts',
+    '.tsx',
+    '.mts',
+    '.cts',
+  ])('preserves externalized %s extensions by default', async (inputExt) => {
     const outdir = join(testRoot, 'out');
     const srcDir = join(testRoot, 'src');
     const stepFile = join(srcDir, 'step.ts');
@@ -80,8 +117,7 @@ describe('createSwcPlugin externalizeNonSteps', () => {
 
     expect(result.errors).toHaveLength(0);
     const output = result.outputFiles[0].text;
-    expect(output).toContain(`/dep${outputExt}`);
-    expect(output).not.toContain(`/dep${inputExt}`);
+    expect(output).toContain(`/dep${inputExt}`);
   });
 });
 

--- a/packages/builders/src/swc-esbuild-plugin.ts
+++ b/packages/builders/src/swc-esbuild-plugin.ts
@@ -19,6 +19,18 @@ export interface SwcPluginOptions {
   outdir?: string;
   projectRoot?: string;
   workflowManifest?: WorkflowManifest;
+  /**
+   * Rewrite TypeScript extensions (.ts, .tsx, .mts, .cts) to their JS
+   * equivalents (.js, .mjs, .cjs) in externalized import paths.
+   *
+   * Enable this when the output bundle is consumed directly by Node's native
+   * ESM loader (e.g. vitest), which cannot resolve .ts extensions.
+   *
+   * Leave disabled (default) when a downstream bundler (webpack, Vite, etc.)
+   * handles resolution — those tools resolve .ts natively and rewriting
+   * breaks them because the .js file doesn't exist on disk.
+   */
+  rewriteTsExtensions?: boolean;
 }
 
 const NODE_RESOLVE_OPTIONS = {
@@ -131,12 +143,14 @@ export function createSwcPlugin(options: SwcPluginOptions): Plugin {
               resolvedPath
             ).replace(/\\/g, '/');
 
-            // Rewrite TypeScript extensions to their JS equivalents so the
-            // externalized import is loadable by Node's native ESM loader.
-            externalPath = externalPath
-              .replace(/\.tsx?$/, '.js')
-              .replace(/\.mts$/, '.mjs')
-              .replace(/\.cts$/, '.cjs');
+            if (options.rewriteTsExtensions) {
+              // Rewrite TypeScript extensions to their JS equivalents so the
+              // externalized import is loadable by Node's native ESM loader.
+              externalPath = externalPath
+                .replace(/\.tsx?$/, '.js')
+                .replace(/\.mts$/, '.mjs')
+                .replace(/\.cts$/, '.cjs');
+            }
           } else {
             externalPath = args.path;
           }

--- a/packages/vitest/src/index.ts
+++ b/packages/vitest/src/index.ts
@@ -50,6 +50,7 @@ class VitestBuilder extends BaseBuilder {
     await this.createStepsBundle({
       outfile: join(this.#outDir, 'steps.mjs'),
       externalizeNonSteps: true,
+      rewriteTsExtensions: true,
       format: 'esm',
       inputFiles,
     });

